pkg/endpointmanager/manager.go

```diff
@@ -72,6 +73,20 @@ type EndpointManager struct {
 
 	// subscribers are notified when events occur in the EndpointManager.
 	subscribers map[Subscriber]struct{}
+
+	// checkHealth supports endpoint garbage collection by verifying the health
+	// of an endpoint.
+	checkHealth EndpointCheckerFunc
+
+	// deleteEndpoint is the function used to remove the endpoint from the
+	// EndpointManager and clean it up. Always set to RemoveEndpoint.
+	deleteEndpoint endpointDeleteFunc
```

Comment:
I haven't fully traced out all the paths of how these are set, but I'm assuming you took the approach of using function pointers instead of interfaces because creating types to implement the interfaces would have been clumsier / more work? I am fine with this approach, so mostly curious as it's not as common as a pattern.

Reply:
Basically yeah. If there was more than one function implemented in one place, I'd figure out the wrapping necessary for having a dedicated structure for it (that's how the original PR did it). In this case, I decided that the implementation of the `EndpointCheckerFunc` should live in `pkg/endpoint` (to avoid pulling a `netlink` dependency and endpoint datapath specifics into `EndpointManager`), then I didn't really want to make it part of the Endpoint as it should be statically callable with an Endpoint parameter, and there's not really an appropriate existing structure to hold this functionality there (unless you get into `EndpointManager`, which is where this ends up attached in practical terms).

For the `endpointDeleteFunc`, I was trying to avoid calling into `Endpoint.Delete()` from the unit tests (for similar reasons, basically the `EndpointManager` shouldn't need to mess with BPF or the filesystem just to validate its own core functionality). In the latter case, I could wrap it in a type... but then what will it be? The `EndpointManagerDeleter`? Certainly doable but I saw no real advantage to such an abstraction.

I also briefly explored abstracting out the `Endpoint` implementation in `pkg/endpointmanager` to deal with abstract endpoints, so that endpointmanager could be tested by itself without hard dependencies on things like the `Endpoint.Delete()` function, but that went down a bit of a rabbithole. In particular it was difficult with the new `Subscriber` interface. Basically if we start abstracting away the `Endpoint` in `EndpointManager` then everything which interacts with the EndpointManager must also deal with the same interface (or assert to concrete types in its own code). I wasn't convinced enough that (1) such abstraction is necessary in the short to medium term, or whether (2) I should also go and refactor that whole set of logic before finally tidying up this PR.

Reply:
Thanks for clarifying. Yeah this makes sense; I agree with wanting to avoid the unneeded abstractions and that they would be too much work for the benefit they'd bring, if any.